File: 91.decode-ways.py
#
# @lc app=leetcode id=91 lang=python3
#
# [91] Decode Ways
#

# @lc code=start
class Solution:
    def numDecodings(self, s: str) -> int:

        s_size = len(s)

        dp = [0 for _ in range(s_size + 1)]

        dp[0] = 1
        dp[1] = 1 if s[0] != "0" else 0

        for i in range(2, s_size + 1):
            if 0 < int(s[i - 1 : i]) <= 9:
                dp[i] += dp[i - 1]
            if 10 <= int(s[i - 2 : i]) <= 26:
                dp[i] += dp[i - 2]
        
        return dp[s_size]
        
# @lc code=end

File: 91.decode-ways.java
/*
 * @lc app=leetcode id=91 lang=java
 *
 * [91] Decode Ways
 */

// @lc code=start
class Solution {
    public int numDecodings(String s) {
        int strLength = s.length();

        int[] dp = new int[strLength + 1];

        dp[0] = 1;
        dp[1] = s.charAt(0) == '0' ? 0 : 1;

        for (int i = 2; i < strLength + 1; i++) {
            
            int oneStep = Integer.valueOf(s.substring(i - 1, i));
            int twoStep = Integer.valueOf(s.substring(i - 2, i));

            if (oneStep >= 1 && oneStep <= 9) {
                dp[i] += dp[i - 1];
            }

            if (twoStep >= 10 && twoStep <= 26) {
                dp[i] += dp[i - 2];
            }
        }

        return dp[strLength];
    }
}
// @lc code=end

